### PR TITLE
ARTEMIS-1194: SOCKS proxy support

### PR DESCRIPTION
**Pull request by adagys**
This work enables Artemis connectors to use a SOCKS proxy, by adding an appropriate netty socks handler to the channel pipeline. The proxy is only enabled for a public remote host.

I tested that it works correctly.

If this is acceptable functionality to be included in Artemis I'm happy to do some more polishing to get it in the right shape.

### CODE REVIEW

**Comment by clebertsuconic** [comment]
nice feature!!!


= it needs a test.. (especially one that uses the URI for the connector)


and... how to test this.. any proxy support in Java we can deploy?

**Comment by jbertram** [comment]
Also needs a JIRA and there are a handful of checkstyle violations.

File: artemis-core-client/src/main/java/org/apache/activemq/artemis/core/remoting/impl/netty/NettyConnector.java [hunk 1]
```diff
@@ -177,6 +181,18 @@
    // will be handled by the server's http server.
    private boolean httpUpgradeEnabled;
 
+   private boolean socksEnabled;
```
Comment (clebertsuconic) [inline artemis-core-client/src/main/java/org/apache/activemq/artemis/core/remoting/impl/netty/NettyConnector.java]: proxyEnabled is a more intuite name... 

especially if you specify a connector with ?useSocket=true;proxyHost=;proxyPort;proxyUser;proxyPassword...

so.. better property names:

- proxyEnabled
- proxyUser
- proxyPassword
- proxyPort
- proxyHost

**Comment by jbertram** [comment]
Please open a JIRA for this, squash all of your commits, and amend your commit message to follow what's outlined in the [Hacking Guide](https://github.com/apache/activemq-artemis/blob/master/docs/hacking-guide/en/code.md#commitMessageDetails).

**Comment by adagys** [comment]
Will do 👍

**Comment by clebertsuconic** [comment]
you could squash all your committs into a single one here:

git rebase -i HEAD~2
# squash them, change the tittle

git push origi -f

**Comment by adagys** [comment]
this is ready for another look @clebertsuconic

**Comment by clebertsuconic** [comment]
I was trying to make a test like this:

https://github.com/clebertsuconic/activemq-artemis/commit/2f7cbc960cc6db4c495425878c9855c84a4c8939


but I'm not sure I will have to time to finish it... if you want to take over this test?

**Comment by adagys** [comment]
can do, so the point of this test is just to check that NettyConnector _can_ establish a connection when the proxy handler is enabled, right?

**Comment by clebertsuconic** [comment]
Yes, basically I wanted to start a proxy, and make sure it worked *through* the proxy.
I was looking to assert the proxy was used.. but I didn't have time to finish the test...

**Comment by clebertsuconic** [comment]
@adagys we still need a way to test it though.. I looked at their website and it seems to support SOCKS.. but I didn't figure out how.

I think we should remove the option you added to ignore localhost though. If it's proxy, always proxy...

**Comment by mtaylor** [comment]
@clebertsuconic  @adagys This has been open since May.  What needs to happen before we can merge/close?

**Comment by clebertsuconic** [comment]
Test

**Comment by adagys** [comment]
I couldn't get it to work in the end and don't have time to work on it anymore.., can be closed

**Comment by clebertsuconic** [comment]
@adagys did you at least try the feature alone? was it working? I couldn't get it to work.

I was interested on the feature.. but I couldn't get the test to work.

**Comment by adagys** [comment]
Yep, I remember seeing it work in an end-to-end scenario, but couldn't get the test to work either.. Which is strange

**Comment by clebertsuconic** [comment]
Can you do one last rebase 👍 
(git pull --rebase upstream master)
... resolve conflicts
git push origin -f



and I will work on the tests soon. (kind of last week)

**Comment by adagys** [comment]
Rebased 👍

**Comment by michaelandrepearce** [comment]
@clebertsuconic whats occuring on this, we ok to merge? And simply mark feature beta until some decent tests?

**Comment by michaelandrepearce** [comment]
@clebertsuconic I'm merging this.

**Comment by michaelandrepearce** [comment]
just needs sorting dependencies, seems the project removed netty all here, so just having to add 

         <dependency>
            <groupId>io.netty</groupId>
            <artifactId>netty-handler-proxy</artifactId>
            <version>${netty.version}</version>
            
         </dependency>
         <dependency>
            <groupId>io.netty</groupId>
            <artifactId>netty-codec-socks</artifactId>
            <version>${netty.version}</version>
            
         </dependency>

**Comment by jbertram** [comment]
@michaelandrepearce, `org.apache.activemq.artemis.tests.unit.core.remoting.impl.netty.NettyConnectorTest#testSocksProxyHandlerAdded` is failing consistently. This may need to be reverted.

**Comment by jbertram** [comment]
@michaelandrepearce, I think I've got a fix for this. Will send a PR shortly...

**Comment by jbertram** [comment]
Just sent #3010.

**Comment by celiovasconcelos** [comment]
```java
    Map<String, Object> params = new HashMap<>();
    params.put(TransportConstants.HOST_PROP_NAME, "localhost");
    params.put(TransportConstants.PORT_PROP_NAME, "61617"); //<acceptor name="netty">tcp://0.0.0.0:61617</acceptor>
    params.put(TransportConstants.PROXY_ENABLED_PROP_NAME, "true");
    params.put(TransportConstants.PROXY_HOST_PROP_NAME, "localhost");
    params.put(TransportConstants.PROXY_PORT_PROP_NAME, "1080");
    params.put(TransportConstants.PROXY_USERNAME_PROP_NAME, "user");
    params.put(TransportConstants.PROXY_PASSWORD_PROP_NAME, "WRONG_PASSWORD");

    TransportConfiguration transportConfiguration = new TransportConfiguration("org.apache.activemq.artemis.core.remoting.impl.netty.NettyConnectorFactory", params);
    ActiveMQConnectionFactory activeMQConnectionFactory = ActiveMQJMSClient.createConnectionFactoryWithoutHA(JMSFactoryType.CF, transportConfiguration);
```
```xml
<dependency>
	<groupId>org.apache.activemq</groupId>
	<artifactId>artemis-jms-client</artifactId>
	<version>2.14.0</version>
</dependency>
```
This is ignoring the proxy conf. 

Am I doing something wrong?

**Comment by jbertram** [comment]
@celiovasconcelos, this is an old, closed pull request. It's not really meant for support. Try using the [ActiveMQ User Mailing List](http://activemq.apache.org/contact/#mailing).

**Comment by celiovasconcelos** [comment]
Sorry. 
I've posted it in https://stackoverflow.com/questions/63248592/amq-artemis-socks-proxy-support
Thank you.

**Pull request closed without merging**